src/coreclr/jit/importer.cpp

```diff
@@ -5278,6 +5270,89 @@ GenTree* Compiler::impArrayAccessIntrinsic(
     }
 }
 
+//------------------------------------------------------------------------
+// impKeepAliveIntrinsic: Import the GC.KeepAlive intrinsic call
+//
+// Imports the intrinsic either as a GT_KEEPALIVE node, or, as an
+// optimization, if the object to keep alive is a GT_BOX, as a series
+// of statements (prepended as needed) with GT_KEEPALIVE top-level nodes,
+// for each field of a reference type in the struct's layout. As part
+// of this, removes the box's allocation and inserts a copy of the
+// box's source to a temporary, if the source wasn't a local already,
+// to enable the re-use of this local for field indirections.
+//
+// Arguments:
+//    objToKeepAlive - the intrinisic call's argument
+//
+// Return Value:
+//    The imported GT_KEEPALIVE, with "objToKeepAlive" as the operand.
+//    If the GT_BOX optimization was successful, always returns a GT_NOP.
+//
+GenTree* Compiler::impKeepAliveIntrinsic(GenTree* objToKeepAlive)
+{
+    assert(objToKeepAlive->TypeIs(TYP_REF));
+
+    if (opts.OptimizationEnabled() && objToKeepAlive->IsBoxedValue())
+    {
+        CORINFO_CLASS_HANDLE boxedClass = lvaGetDesc(objToKeepAlive->AsBox()->BoxOp()->AsLclVar())->lvClassHnd;
+        GenTree*             boxSrc = gtTryRemoveBoxUpstreamEffects(objToKeepAlive, Compiler::BR_REMOVE_BUT_NOT_NARROW);
+
+        if (boxSrc != nullptr)
+        {
+            JITDUMP("\nImporting KEEPALIVE(BOX) as a series of KEEPALIVE(FIELD)\n");
+            INDEBUG(bool foundGcField = false);
+
+            unsigned boxTempNum;
+            if (boxSrc->OperIsLocal())
+            {
+                boxTempNum = boxSrc->AsLclVarCommon()->GetLclNum();
+            }
+            else
+            {
+                boxTempNum            = lvaGrabTemp(true DEBUGARG("Temp for the box source"));
+                GenTree*   boxTempAsg = gtNewTempAssign(boxTempNum, boxSrc);
+                Statement* boxAsgStmt = objToKeepAlive->AsBox()->gtCopyStmtWhenInlinedBoxValue;
+                boxAsgStmt->SetRootNode(boxTempAsg);
+            }
+
+            ClassLayout* layout = typGetObjLayout(boxedClass);
+            if (layout->HasGCPtr())
+            {
+                for (unsigned slot = 0; slot < layout->GetSlotCount(); slot++)
+                {
+                    if (layout->IsGCPtr(slot))
+                    {
+                        // This is a very verbose way of saying gtNewLclFldNode.
```

Comment:
why importation was chosen for this transformation? Won't it work better after implicit byrefs resolution somewhere in morph? Are these `IND(ADD(ADDR(LCL_VAR), OFFSET) ` always transformed as `GT_LCL_FLD`?

Reply:
> why importation was chosen for this transformation?

Primary reason is because that's where the original code was. I can think of a few reasons that make it beneficial to do in import:
1) Adding statements is "natural", and `KEEPALIVE`s require new statements (except if one uses `COMMA`s I suppose, but I'd think it is better not to use them if we can).
2) Boxes are "fresh", i. e. there is little chance that some subsequent transform changed their shape somehow.
3) Implicit byrefs and normal locals are "the same" (no need to write code that handles both - though perhaps that code is trivial enough for this not to matter).

However, I am in now way as experienced as you in this area. What would be the advantage of doing it late(er)?

> Are these `IND(ADD(ADDR(LCL_VAR), OFFSET)` always transformed as `GT_LCL_FLD`?

For locals, yes. I expect most of these structs will be implicit byrefs though, for them it turns into `IND(ADD(LCL_VAR byref, offset))`.

Reply:
> Implicit byrefs and normal locals are "the same" (no need to write code that handles both - though perhaps that code is trivial enough for this not to matter).

for implicit byrefs do we need to do anything? are not they reported as alive from the parent method, for example:
```
struct S // more than 16 bytes, so implicit byref
{
 long a, b, c;
}

void Foo()
{
  S s = new S();
  Do(s); // <- implicit byref, s is on the stack and reported live for this line
}

void Do(S s) // <- s is implicit byref so we don't care about its liveness because the caller is reporting it.
{
  CallUnmanagedApiWithSuppressGCTransition(s); // no need to report s as live in Do, it will be reported in Foo.
}
```

Reply:
> for implicit byrefs do we need to do anything?

Sorry for not being explicit. I agree there are liveness concerns here, it is just convenient that the implicit byrefs haven't been expanded yet, so the copy elision code (the `if (OperIsLocal()) { reuse the existing local }` branch) "just works".

On that note this copy elision is actually bugged it must be `if (OperIs(GT_LCL_VAR))`.